components/script/dom/node.rs

```diff
@@ -1033,6 +1107,27 @@ impl Node {
             None
         }
     }
+
+    /// https://dom.spec.whatwg.org/#retarget
+    pub fn retarget(&self, b: &Node) -> DomRoot<Node> {
+        let mut a = DomRoot::from_ref(&*self);
+        loop {
+            // Step 1.
+            let a_root = a.GetRootNode(&GetRootNodeOptions::empty());
```

Comment:
You should make sure to optimize `GetRootNode` in a shadow tree to be fast and return the containing shadow, and similar for nodes in the document to return the document.